lib/utils/aws/signing.go

```diff
@@ -0,0 +1,194 @@
+	// GetSigningCredentials allows to set the function responsible for obtaining STS credentials.
+	// Used in tests to set static AWS credentials and skip API call.
+	GetSigningCredentials GetSigningCredentialsFunc
+}
+
+// CheckAndSetDefaults validates the SigningServiceConfig config.
+func (s *SigningServiceConfig) CheckAndSetDefaults() error {
+	if s.Clock == nil {
+		s.Clock = clockwork.NewRealClock()
+	}
+	if s.Session == nil {
+		ses, err := awssession.NewSessionWithOptions(awssession.Options{
+			SharedConfigState: awssession.SharedConfigEnable,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s.Session = ses
+	}
+	if s.GetSigningCredentials == nil {
+		s.GetSigningCredentials = GetAWSCredentialsFromSTSAPI
+	}
+	return nil
+}
+
+// SigningCtx contains AWS SigV4 signing context parameters.
+type SigningCtx struct {
+	SigningName   string
+	SigningRegion string
+	Expiry        time.Time
+	SessionName   string
+	AWSRoleArn    string
+	AWSExternalID string
+}
+
+// Check checks signing context parameters.
+func (sc *SigningCtx) Check(clock clockwork.Clock) error {
+	switch {
+	case sc.SigningName == "":
+		return trace.BadParameter("missing AWS signing name")
+	case sc.SigningRegion == "":
+		return trace.BadParameter("missing AWS signing region")
+	case sc.SessionName == "":
+		return trace.BadParameter("missing AWS session name")
+	case sc.AWSRoleArn == "":
+		return trace.BadParameter("missing AWS Role ARN")
+	case sc.Expiry.Before(clock.Now()):
+		return trace.BadParameter("AWS SigV4 expiry has already expired")
+	default:
+		return nil
+	}
+}
+
+// SignRequest creates a new HTTP request and rewrites the header from the original request and returns a new
+// HTTP request signed by STS AWS API.
+// Signing steps:
+// 1) Decode Authorization Header. Authorization Header example:
+//
+//		Authorization: AWS4-HMAC-SHA256
+//		Credential=AKIAIOSFODNN7EXAMPLE/20130524/us-east-1/s3/aws4_request,
+//		SignedHeaders=host;range;x-amz-date,
+//		Signature=fe5f80f77d5fa3beca038a248ff027d0445342fe2855ddc963176630326f1024
+//
+//	 2. Extract credential section from credential Authorization Header.
+//	 3. Extract aws-region and aws-service from the credential section.
+//	 4. Build AWS API endpoint based on extracted aws-region and aws-service fields.
+//	    Not that for endpoint resolving the https://github.com/aws/aws-sdk-go/aws/endpoints/endpoints.go
+//	    package is used and when Amazon releases a new API the dependency update is needed.
+//	 5. Sign HTTP request.
+func (s *SigningService) SignRequest(req *http.Request, signCtx *SigningCtx) (*http.Request, error) {
+	if signCtx == nil {
+		return nil, trace.BadParameter("missing signing context")
+	}
+	if err := signCtx.Check(s.Clock); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	payload, err := GetAndReplaceReqBody(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	reqCopy, err := http.NewRequest(req.Method, req.URL.String(), bytes.NewReader(payload))
```

Comment:
`req.Clone()` followed by `req.Body = ...`

Reply:
I don't think the `req.Body` change is needed since we called `GetAndReplaceReqBody(req)` already?

Reply:
nvm I think it's safer to just assign to body after cloning